[Draft] Ncs thermostat sample

LipinskiPNordicSemi opened the pull request:
- [x]     implement the sample thermostat application according to the agreed architecture
- [x]     prepare documentation of the sample (keep the style and docs structure from other NCS Matter samples)
- [x]     make sure the DFU works (both BLE SMP and Matter OTA), to be tested with mcumgr and OTA Provider accordingly
- [x]     add a possibility to trigger BLE advertising by pressing a button (consider that on nRF7002 we have only 2 buttons)
- [x]     consider adding Kconfigs for simulated temperature measurement (e.g. simulated temperature range, random generator refresh interval, increment step etc.)
- [x]     consider definition of template-based temperature sensor interface (reusable code for both simulated and real temperature measurement - nRF52/53 DKs vs. Thingy:53)

markaj-nordic commented on samples/matter/light_bulb/src/app_task.cpp:
```diff
@@ -659,4 +659,4 @@ void AppTask::UpdateClusterState()
 		}
 	});
-}
+}
```
New line missing at the end of the file

LipinskiPNordicSemi replied:
By mistake, I've edited this file and saved. I had to copy the whole file from github

markaj-nordic commented on samples/matter/lock/src/app_task.cpp:
```diff
@@ -760,4 +760,4 @@ void AppTask::NUSUnlockCallback(void *context)
 	}
 }
-#endif
+#endif
```
New line missing at the end of the file

markaj-nordic replied:
Why was this file modified anyway?

LipinskiPNordicSemi replied:
By mistake, I've edited this file and saved. I had to copy the whole file from github

markaj-nordic replied:
Still, looks like the new line has been removed from this file by mistake.

markaj-nordic commented on samples/matter/thermostat/CMakeLists.txt:
```diff
@@ -0,0 +1,73 @@
+    target_sources(app PRIVATE ${COMMON_ROOT}/src/ota_util.cpp)
+endif()
+
+if(CONFIG_CHIP_NUS)
```
NUS shall not be supported in this sample at all.

markaj-nordic commented on samples/matter/thermostat/Kconfig:
```diff
@@ -0,0 +1,49 @@
+	default y
+
+config TEMPERATURE_STEP
+	int "If this config is se to to 0, mocked value will be random. Mocked temperature steps. 10 equels 0,1 degrees."
```
```suggestion
	int "If this config is set to to 0, mocked value will be random. Mocked temperature steps. 10 equals 0,1 degrees."
```

markaj-nordic replied:
In general we should rephrase it somehow.

markaj-nordic commented on samples/matter/thermostat/Kconfig:
```diff
@@ -0,0 +1,49 @@
+	default 0
+
+config MAXIMUM_MOCKED_TEMPERATURE
+	int "Max value in mocked temperature range"
```
Let's not using shortcuts, line min/max in the Kconfig description.

markaj-nordic commented on samples/matter/thermostat/Kconfig:
```diff
@@ -0,0 +1,49 @@
+endchoice
+
+endif # NET_L2_OPENTHREAD
+
```
To many empty lines

markaj-nordic commented on samples/matter/thermostat/README.rst:
```diff
@@ -0,0 +1,224 @@
+       - Manual pairing code
+     * - Scan the following QR code with the app for your ecosystem:
+
+         .. figure:: ../../../doc/nrf/images/matter_qr_code_template_sample.png
```
I guess we should generate dedicated QR code for the Thermostat sample.

markaj-nordic replied:
@LipinskiPNordicSemi do we have a dedicated QR code now?

markaj-nordic commented on samples/matter/thermostat/README.rst:
```diff
@@ -0,0 +1,224 @@
+.. _matter_template_sample:
+
+Matter: Template
```
```suggestion
Matter: Thermostat
```
In general this file needs to be adapted to the Thermostat sample.

markaj-nordic commented on samples/matter/thermostat/README.rst:
```diff
@@ -0,0 +1,224 @@
+
+      File not found: ./ncs/nrf/samples/matter/template/configuration/nrf52840dk_nrf52840/prj_shell.conf
+
+Testing
```
In this section we need to add the proper testing scenario for the thermostat, including description of the setup with external source of temperature measurement (binding).

markaj-nordic replied:
I mean, we need to clearly describe that the thermostat can operate in mocked temperature measurement mode or collect the data remotely from bound device. This section should explain the usage of defined Kconfigs and refer to the section with binding described. We should also briefly what binding actually means from Matter perspective. Please take a look at the light switch documentation, some section may be taken from there I guess:
https://developer.nordicsemi.com/nRF_Connect_SDK/doc/latest/nrf/samples/matter/light_switch/README.html#binding

markaj-nordic commented on samples/matter/thermostat/prj.conf:
```diff
@@ -0,0 +1,46 @@
+#
+# Copyright (c) 2021 Nordic Semiconductor ASA
```
Update the year, make sure it's proper in all files.

markaj-nordic commented on samples/matter/thermostat/src/app_config.h:
```diff
@@ -0,0 +1,36 @@
+
+#pragma once
+
+/* ---- Template App Config ---- */
```
Thermostat

markaj-nordic commented on samples/matter/thermostat/src/app_config.h:
```diff
@@ -0,0 +1,36 @@
+
+#include "board_util.h"
+
+#define SENSOR_MOCK 1
```
Is is used anywhere?

markaj-nordic commented on samples/matter/thermostat/src/app_task.cpp:
```diff
@@ -0,0 +1,549 @@
+using namespace ::chip::app;
+using namespace ::chip::Credentials;
+using namespace ::chip::DeviceLayer;
+
```
too many lines

markaj-nordic commented on samples/matter/thermostat/src/app_task.cpp:
```diff
@@ -0,0 +1,549 @@
+        return err;
+    }
+
+    err = TemperatureManager().Init();
```
Why do we init the temperature manager twice?

markaj-nordic commented on samples/matter/thermostat/src/temp_sensor_manager.cpp:
```diff
@@ -0,0 +1,142 @@
+k_timer sSensorTimer;
+
+#if CONFIG_TEMPERATURE_STEP == 0
+static int16_t mMockTemp[]               = {2000, 
```
Make it const and use `s` prefix instead of `m`.

markaj-nordic commented on samples/matter/thermostat/src/temp_sensor_manager.cpp:
```diff
@@ -0,0 +1,142 @@
+                                            2999};
+
+#else
+static int16_t mMockedTemperatureStep = CONFIG_TEMPERATURE_STEP;
```
Use `constexpr` for compile time consts.
Also, `m` prefix is dedicated for internal class members. Here, please use `k` prefix.

markaj-nordic commented on samples/matter/thermostat/src/app_task.cpp:
```diff
@@ -0,0 +1,549 @@
+}
+
+
+bool sIsNetworkProvisioned = false;
```
statics

markaj-nordic commented on samples/matter/thermostat/src/temperature_manager.h:
```diff
@@ -0,0 +1,58 @@
+
+private:
+
+    uint8_t ReturnCompleteValue(int16_t Value);
```
These are stateless helpers, make it static and maybe even move outside of TemperatureMeanager namespace (static functions in the .cpp file)

markaj-nordic commented on samples/matter/thermostat/src/app_task.h:
```diff
@@ -0,0 +1,78 @@
+	static void LEDStateUpdateHandler(LEDWidget &ledWidget);
+	static void FunctionTimerTimeoutCallback(k_timer *timer);
+	static void UpdateStatusLED();
+
```
empty lines

markaj-nordic commented on samples/matter/thermostat/src/temp_sensor_manager.cpp:
```diff
@@ -0,0 +1,142 @@
+{
+    // Initialize temp sensor timer
+
+    int16_t InitValue            = 0;
```
In really doesn't matter, but please keep the `camelCase` for temporary variables as well.

markaj-nordic commented on samples/matter/thermostat/src/temp_sensor_manager.cpp:
```diff
@@ -0,0 +1,142 @@
+    int16_t InitValue            = 0;
+    
+    PlatformMgr().LockChipStack();
+    app::Clusters::Thermostat::Attributes::LocalTemperature::Set(kThermostatEndpoint, InitValue);
```
Shouldn't it be configured in the ZAP file? Anyway, if we want to modify it at init, better do it in the emberAfThermostatClusterInitCallback (should be done in the zcl_callbacks.cpp). I guess, we don't need to lock the chip stack then.

LipinskiPNordicSemi replied:
I'm not able to set any value in the zap tool as default for LocalTemperature:
![image](https://github.com/LipinskiPNordicSemi/sdk-nrf/assets/132913157/9bf3651d-5adc-4b1f-bb51-d60d815f8152)

markaj-nordic replied:
But it can be set to NULL, right?

markaj-nordic commented on samples/matter/thermostat/src/temp_sensor_manager.cpp:
```diff
@@ -0,0 +1,142 @@
+    app::Clusters::Thermostat::Attributes::LocalTemperature::Set(kThermostatEndpoint, InitValue);
+    PlatformMgr().UnlockChipStack();
+
+    AppEvent event;
```
I think, ideally, we should only post events in the app_task module and leave all other modules as much independent from app_task as possible.

ArekBalysNordic replied:
Do we need to run it from the App task context?

LipinskiPNordicSemi replied:
II think it would be problematic to move this section to app task (making two handlers public for TempSensorManager class and app_task class). If there is a better option, I can't see it

markaj-nordic replied:
I bet this is doable by providing a proper API. But let's leave it for now. Maybe I will try to address this while developing a proposal for the new generic API for our samples, which will land in our backlog.

markaj-nordic commented on samples/matter/thermostat/src/temp_sensor_manager.cpp:
```diff
@@ -0,0 +1,142 @@
+    
+    #if CONFIG_TEMPERATURE_STEP == 0
+    
+    int16_t Temperature            = 0;
```
can be initialized with the `mMockTemp[MockTempIndex]`.

markaj-nordic commented on samples/matter/thermostat/src/temp_sensor_manager.cpp:
```diff
@@ -0,0 +1,142 @@
+    #ifdef CONFIG_TEMPERATURE_SENSORS_MOCKED
+
+
+    static uint8_t MockTempIndex = 0;
```
Again, I prefer to keep `camelCase` everywhere if possible.

markaj-nordic commented on samples/matter/thermostat/src/temp_sensor_manager.cpp:
```diff
@@ -0,0 +1,142 @@
+{
+    static int16_t PreviousTemperature = 0;
+
+    #ifdef CONFIG_TEMPERATURE_SENSORS_MOCKED
```
#ifdefs should be aligned to the left side

markaj-nordic commented on samples/matter/thermostat/src/temp_sensor_manager.h:
```diff
@@ -0,0 +1,35 @@
+    private:
+        static void TimerEventHandler(k_timer * timer);
+        static void SensorTimerEventHandler(const AppEvent &Event);
+};
```
Missing new line at the end of file

markaj-nordic commented on samples/matter/thermostat/src/temp_sensor_manager.h:
```diff
@@ -0,0 +1,35 @@
+            return sTempSensorManager;
+        };
+
+        void TemperatureChangeHandler(AttributeId attributeId, uint8_t * value, uint16_t size);
```
Is it Github parsing problem or there is really no definition of this method?

LipinskiPNordicSemi replied:
I changed the idea in the middle of development, and didn't realise it is still in the code.

markaj-nordic commented on samples/matter/thermostat/src/temp_sensor_manager.cpp:
```diff
@@ -0,0 +1,142 @@
+    CycleCounter++;
+    if (CycleCounter >= kSimulatedReadingFrequency)
+    {
+        if(MockTempIndex >= ArraySize(mMockTemp) - 1){
```
Is this condition correct? 
```suggestion
        if(MockTempIndex > ArraySize(mMockTemp) - 1){
```

markaj-nordic commented on samples/matter/thermostat/src/temp_sensor_manager.cpp:
```diff
@@ -0,0 +1,142 @@
+        CycleCounter = 0;
+    }
+    #else 
+        #ifdef CONFIG_MOCKED_TEMPERATURE_DIRRECTION
```
This should be configured purely in the preprocessor:
#ifdef CONFIG_MOCKED_TEMPERATURE_DIRRECTION
#define INITIAL_TEMPERATURE CONFIG_MINIMUM_MOCKED_TEMPERATURE
#else
#define INITIAL_TEMPERATURE CONFIG_MAXIMUM_MOCKED_TEMPERATURE
...
static int16_t Temperature = INITIAL_TEMPERATURE;

Anyway, I don't quite get this code, do we even have MOCKED_TEMPERATURE_DIRRECTION Kconfig defined anywhere?

Also, typo in the MOCKED_TEMPERATURE_DIRRECTION -> MOCKED_TEMPERATURE_DIRECTION

LipinskiPNordicSemi replied:
I changed the config name and forgot to change it in the code

markaj-nordic commented on samples/matter/thermostat/src/temp_sensor_manager.cpp:
```diff
@@ -0,0 +1,142 @@
+    {
+        PreviousTemperature = Temperature;
+        PlatformMgr().LockChipStack();
+        app::Clusters::Thermostat::Attributes::LocalTemperature::Set(kThermostatEndpoint, Temperature);
```
I'm not really sure if we need to manipulate `LocalTemperature` value. In our use cases, we either emulate the temperature with some dummy data or remotely get the measurements from the real temperature sensor device. Is the `LocalTemperature` attribute mandatory for Thermostat cluster?

LipinskiPNordicSemi replied:
It is Mandatory for thermostat cluster.

markaj-nordic commented on samples/matter/thermostat/src/temperature_manager.cpp:
```diff
@@ -0,0 +1,137 @@
+CHIP_ERROR TemperatureManager::Init()
+{
+    app::DataModel::Nullable<int16_t> temp;
+    int16_t heatingSetpoint, coolingSetpoint;
```
seems redundant, we can write these values directly to the class members, e.g. `&mHeatingCelsiusSetPoint`

markaj-nordic commented on samples/matter/thermostat/src/temperature_manager.cpp:
```diff
@@ -0,0 +1,137 @@
+    PlatformMgr().UnlockChipStack();;
+
+
+    Instance().mCurrentTempCelsius     = temp.Value();
```
Init() is a non-static method, so we don't need to utilize Instance() to access class members

markaj-nordic commented on samples/matter/thermostat/src/app_task.cpp:
```diff
@@ -0,0 +1,549 @@
+	PostEvent(event);
+}
+
+void AppTask::UpdateThermostatStatus(void)
```
It's not really updating anything, it just logs the thermostat status, so either change the naming or just use `TemperatureManager::Instance().LogThermostatstatus();` directly. when needed.

markaj-nordic commented on samples/matter/thermostat/src/temperature_manager.cpp:
```diff
@@ -0,0 +1,137 @@
+uint8_t TemperatureManager::ReturnReminderValue(int16_t Value)
+{
+
+    uint8_t remainder = (Value%100 + 5)/10;
```
I suggest to add a comment describing this formula.

markaj-nordic commented on samples/matter/thermostat/src/temperature_manager.cpp:
```diff
@@ -0,0 +1,137 @@
+    switch (attributeId)
+    {
+    case ThermostatAttributes::LocalTemperature::Id: {
+        int16_t Temp = (*((int16_t *) value));
```
If any casting is needed please use C++ style `static_cast` or any other similar

markaj-nordic commented on samples/matter/thermostat/src/temperature_manager.h:
```diff
@@ -0,0 +1,58 @@
+using namespace chip;
+
+// AppCluster Spec Table 85.
+typedef enum
```
We don't need typedef for enums in C++.

markaj-nordic reviewed (commented):
Added an initial review. One general comment is to format the whole code with clang-format.

Damian-Nordic commented on samples/matter/thermostat/CMakeLists.txt:
```diff
@@ -0,0 +1,72 @@
+
+cmake_minimum_required(VERSION 3.20.0)
+
+get_filename_component(MATTER_MODULE_ROOT $ENV{ZEPHYR_BASE}/../modules/lib/matter REALPATH)
```
Take a look at the newest version of Matter samples in NCS - this line has been removed to remove the dependency on ZEPHYR_BASE env variable.

Damian-Nordic commented on samples/matter/thermostat/src/temperature_measurement/real_temp.h:
```diff
@@ -0,0 +1,29 @@
+
+class RealSensor 
+{
+  public:
```
Please install "clang-format" and integrate it in VS Code (or whatever editor you use) so that you automatically apply the style defined in `samples/matter/.clang-format` and save your time on code formatting :)

ArekBalysNordic commented on samples/matter/thermostat/Kconfig:
```diff
@@ -0,0 +1,52 @@
+
+config TEMPERATURE_SENSORS_MOCKED
+	bool "Define if the temperature is mocked or if it is given by internal setup sensors[only for Thingy53]"
+	default y
```
I guess we can merge these two configs together. If the REAL_SENSOR is set to n then we will use the mocked one. Second thing, It is worth adding some more name descriptions to this config, for example, TERMOSTAT_REAL_SENSOR. And the third thing, configs always are by default set to n, so it is better to not write `default n` in such cases.

ArekBalysNordic replied:
Please consider changing names also in other configs.

ArekBalysNordic commented on samples/matter/thermostat/Kconfig:
```diff
@@ -0,0 +1,52 @@
+	int "Defines mocekd temperature step. If 0 then temperature is random"
+	default 0
+	help 
+		Defines temperature step, which will be used to define differeces between next mocekd temperature values.
```
Wrong indentation of help description, please see somewhere else how it should look like.

kkasperczyk-no replied:
Typos
```suggestion
		Defines temperature step, which will be used to define differences between next mocked temperature values.
```

ArekBalysNordic commented on samples/matter/thermostat/Kconfig:
```diff
@@ -0,0 +1,52 @@
+	default 1500
+
+config MOCKED_TEMPERATURE_RAISE
+	bool "If this option is set to 'true' temperature will go from min to max, if set to 'false' temperature will go from max to min"
```
The config description should be as short as possible because it will be cut in Zephyr's config creator. Please use the help field to describe the functionality precisely.

ArekBalysNordic commented on samples/matter/thermostat/Kconfig:
```diff
@@ -0,0 +1,52 @@
+#
+# SPDX-License-Identifier: LicenseRef-Nordic-5-Clause
+#
+mainmenu "Matter nRF Connect Thermostat Example Application"
```
```suggestion
mainmenu "Matter Thermostat sample application"
```

ArekBalysNordic commented on samples/matter/thermostat/boards/nrf52840dk_nrf52840.overlay:
```diff
@@ -0,0 +1,66 @@
+	};
+
+	soc{
+		temp: temp@4000c000 {
```
I heard recently that you resigned from measuring SoC temperature, is it still needed?

ArekBalysNordic commented on samples/matter/thermostat/boards/nrf52840dk_nrf52840.overlay:
```diff
@@ -0,0 +1,66 @@
+};
+
+&nvic {
+	arm,num-irq-priority-bits = <3>;
```
Why is it needed?

ArekBalysNordic commented on samples/matter/thermostat/prj.conf:
```diff
@@ -0,0 +1,50 @@
+
+CONFIG_SENSOR=y
+
+CONFIG_TEMP_NRF5=y
```
Missing a new line at the end of the file.

ArekBalysNordic commented on samples/matter/thermostat/sample.yaml:
```diff
@@ -0,0 +1,36 @@
+    build_only: true
+    integration_platforms:
+      - nrf7002dk_nrf5340_cpuapp
+    platform_allow: nrf7002dk_nrf5340_cpuapp
```
Please change everywhere `Template` it to `Thermostat`, or something like that :)

ArekBalysNordic commented on samples/matter/thermostat/child_image/multiprotocol_rpmsg/prj_no_dfu.conf:
```diff
@@ -0,0 +1,15 @@
+#
+# Copyright (c) 2022 Nordic Semiconductor ASA
```
Please set 2023 year in each new file.

ArekBalysNordic commented on samples/matter/thermostat/src/app_task.cpp:
```diff
@@ -0,0 +1,542 @@
+	err = TempSensorManager().Init();
+	if (err != CHIP_NO_ERROR)
+    {
+        LOG_ERR("SensorMgr Init fail");
```
Why log says `SensorMgr` when you've called a class above `TempSensorManager`?

ArekBalysNordic commented on samples/matter/thermostat/src/app_task.cpp:
```diff
@@ -0,0 +1,542 @@
+
+void AppTask::ThermostatHandler(const AppEvent &event)
+{
+
```
```suggestion
```

ArekBalysNordic commented on samples/matter/thermostat/src/binding_handler.cpp:
```diff
@@ -0,0 +1,155 @@
+
+void BindingHandler::Init()
+{
+
```
```suggestion
```

ArekBalysNordic commented on samples/matter/thermostat/src/binding_handler.cpp:
```diff
@@ -0,0 +1,155 @@
+
+	
+	TemperatureMeasurementUnicastBindingRead(data, binding, peer_device);
+
```
```suggestion
```

ArekBalysNordic commented on samples/matter/thermostat/src/binding_handler.cpp:
```diff
@@ -0,0 +1,155 @@
+{
+	VerifyOrReturn(context != nullptr, LOG_ERR("Invalid context for Thermostat handler"););
+	BindingData *data = static_cast<BindingData *>(context);
+
```
```suggestion
```

ArekBalysNordic commented on samples/matter/thermostat/src/binding_handler.h:
```diff
@@ -0,0 +1,73 @@
+#include <app/clusters/bindings/BindingManager.h>
+#include <app/OperationalSessionSetup.h>
+#include "controller/InvokeInteraction.h"
+#include "controller/ReadInteraction.h"
```
```suggestion
#include <controller/InvokeInteraction.h>
#include <controller/ReadInteraction.h>
```

ArekBalysNordic commented on samples/matter/thermostat/src/chip_project_config.h:
```diff
@@ -0,0 +1,16 @@
+ *
+ */
+
+#pragma once
```
Hmm, nothing is needed here?

ArekBalysNordic replied:
.

LipinskiPNordicSemi replied:
I think it should stay empty

markaj-nordic replied:
Yes, if there is no need for specific CHIP configuration this should remain empty (e.g. for window covering sample it's the same)

ArekBalysNordic commented on samples/matter/thermostat/src/temperature_measurement/real_temp.h:
```diff
@@ -0,0 +1,29 @@
+ *
+ * SPDX-License-Identifier: LicenseRef-Nordic-5-Clause
+ */
+#include <stdint.h>
```
Lack of `#pragma once`. Why do you need separate .h files for real_temp and mocked_temp? It looks like both solutions have the same API.

ArekBalysNordic commented on samples/matter/thermostat/src/temperature_measurement/real_temp.cpp:
```diff
@@ -0,0 +1,48 @@
+
+namespace{
+
+constexpr int16_t kTemperatureMeasurementAttributeMaxValue = 0x7fff;
+constexpr int16_t kTemperatureMeasurementAttributeMinValue = 0x954d;
+constexpr int16_t kTemperatureMeasurementAttributeInvalidValue = 0x8000;
```
I wonder whether we have those numbers defined somewhere in Matter stack / ZCL.

markaj-nordic replied:
@LipinskiPNordicSemi, do we?

LipinskiPNordicSemi replied:
I'm not able to find anything similar except weather station where it is declared also

ArekBalysNordic replied:
So, from where do you get those values? Because they are quite unusual and very specific.

markaj-nordic replied:
I presume these must come from the TemperatureMeasurement attribute specification, I agree that we should at least comment to point out the source of this data.

ArekBalysNordic commented on samples/matter/thermostat/src/temperature_measurement/mocked_temp.cpp:
```diff
@@ -0,0 +1,76 @@
+#if CONFIG_TEMPERATURE_STEP == 0
+    
+    static uint8_t mockTempIndex    = 0;
+    static uint8_t cycleCounter     = 0;
```
What do you thing about making those variables class members?

ArekBalysNordic commented on samples/matter/thermostat/src/temp_sensor_manager.h:
```diff
@@ -0,0 +1,40 @@
+
+using namespace chip;
+
+constexpr uint16_t kSensorTimerPeriodMs  = 30000; // 30s timer period
```
Move it to class?

kkasperczyk-no replied:
I'm wondering if we need a class in this file at all :) This is a singleton object with all static methods and no fields to set, so I don't see a benefit from using a class compared to just implementing the methods in the C-style.

kkasperczyk-no replied:
Another thing is that `MockedSensor` and `RealSensor` are also pure static class objects. I'm not saying it's wrong, but yeah it's kind of unnecessary to me.

ArekBalysNordic commented on samples/matter/thermostat/src/temp_sensor_manager.cpp:
```diff
@@ -0,0 +1,95 @@
+void TempSensorManager::StartMeasurement(SensorType const& Sensor)
+{
+
+    Sensor.Instance().Init();
```
Do you need to call Init() method on each StartMeasurement invocation?

ArekBalysNordic commented on samples/matter/thermostat/src/temp_sensor_manager.cpp:
```diff
@@ -0,0 +1,95 @@
+
+    k_timer_init(&sSensorTimer, &TempSensorManager::TimerEventHandler, nullptr);
+    k_timer_user_data_set(&sSensorTimer, this);
+    k_timer_start(&sSensorTimer, K_MSEC(kSensorTimerPeriodMs), K_NO_WAIT);
```
You start the timer here and in `SensorTimerEventHandler` method. Do you think it is correct? When the timer is initialized?

kkasperczyk-no replied:
Yeah, architecture here is a little bit confusing, as we post an event in this method, which results in starting the timer and the timer handler posts an event again. On the other hand we start the timer here, which handler posts an event.

I think it would generally simplify the implementation, if you would change starting timer with `K_NO_WAIT` to the `K_MSEC(kSensorTimerPeriodMs)`. Now you re-start the timer every time, and then you would just start it once, and the following handlers would be called every `kSensorTimerPeriodMs` automatically.

ArekBalysNordic commented on samples/matter/thermostat/src/temperature_manager.h:
```diff
@@ -0,0 +1,58 @@
+}SystemMode;
+
+
+static uint8_t sReturnCompleteValue(int16_t Value);
+static uint8_t sReturnReminderValue(int16_t Value);
```
I guess you tried to declare ReturnReminderValue & ReturnCompleteValue functions here, but `s` has gone away. Moreover, I guess we should stick C++ style in our samples. Follow up on that in the comment regarding the usage of these functions.

LipinskiPNordicSemi replied:
That might be misunderstood of previous comment of @markaj-nordic 
![image](https://github.com/LipinskiPNordicSemi/sdk-nrf/assets/132913157/5c45d8dc-1f8c-4ef7-b30f-6af2865e5783)

ArekBalysNordic commented on samples/matter/thermostat/src/temperature_manager.cpp:
```diff
@@ -0,0 +1,134 @@
+{
+
+    LOG_INF("Thermostat Status - M:%d T:%d,%d'C H:%d,%d'C C:%d,%d'C", GetMode(), 
+                                                            ReturnCompleteValue(GetCurrentTemp()),  ReturnReminderValue(GetCurrentTemp()),
```
To stick c++ style you can declare ReturnReminderValue & ReturnCompleteValue only within this file by declaring them as `static` or putting them in an anonymous namespace, or just create lambdas here since they are used only in this method.

ArekBalysNordic commented on samples/matter/thermostat/src/temperature_manager.cpp:
```diff
@@ -0,0 +1,134 @@
+void TemperatureManager::LogThermostatStatus()
+{
+
+    LOG_INF("Thermostat Status - M:%d T:%d,%d'C H:%d,%d'C C:%d,%d'C", GetMode(), 
```
For me, this log is pretty hard to understand, can you describe to me the Reminder value?

ArekBalysNordic reviewed (commented):
The first round of comments.

kkasperczyk-no commented on samples/matter/thermostat/Kconfig:
```diff
@@ -0,0 +1,52 @@
+	default y
+
+config TEMPERATURE_STEP
+	int "Defines mocekd temperature step. If 0 then temperature is random"
```
Typo
```suggestion
	int "Defines mocked temperature step. If 0 then temperature is random"
```

kkasperczyk-no commented on samples/matter/thermostat/prj.conf:
```diff
@@ -0,0 +1,50 @@
+# 32768 == 0x8000 (example Product ID added temporaly,
+# but it must be changed with proper PID from the list:
+# https://github.com/project-chip/connectedhomeip/blob/482e6fd03196a6de45465a90003947ef4b86e0b1/docs/examples/discussion/PID_allocation_for_example_apps.md)
+CONFIG_CHIP_DEVICE_PRODUCT_ID=32768
```
0x8000 was a correct value for the template sample, however following the link above, the proper value for thermostat is 0x800e (32782)

kkasperczyk-no commented on samples/matter/thermostat/src/temp_sensor_manager.cpp:
```diff
@@ -0,0 +1,95 @@
+    int16_t initValue            = 0;
+    
+    PlatformMgr().LockChipStack();
+    app::Clusters::Thermostat::Attributes::LocalTemperature::Set(kThermostatEndpoint, initValue);
```
I think you could call `SetLocalTemperature` here instead of duplicating the code.

kkasperczyk-no commented on samples/matter/thermostat/src/temp_sensor_manager.cpp:
```diff
@@ -0,0 +1,95 @@
+#include "app_config.h"
+#include "app_task.h"
+#include "app_event.h"
+#include "temperature_measurement/real_temp.h"
```
Considering includes done few lines below, I guess these two should be removed.

kkasperczyk-no commented on samples/matter/thermostat/src/temp_sensor_manager.cpp:
```diff
@@ -0,0 +1,95 @@
+    AppTask::Instance().PostEvent(event);
+
+    k_timer_init(&sSensorTimer, &TempSensorManager::TimerEventHandler, nullptr);
+    k_timer_user_data_set(&sSensorTimer, this);
```
Minor: If you don't need to use the user data in the timer handler (which we don't), setting user data pointer to `this` instance is not necessary.

kkasperczyk-no commented on samples/matter/thermostat/src/temp_sensor_manager.cpp:
```diff
@@ -0,0 +1,95 @@
+void TempSensorManager::SensorTimerEventHandler(const AppEvent &Event)
+{
+    
+    StartMeasurement(RealSensor::Instance()); // Initiate measurement, by giving RealSensor instance  function read 
```
Please use /* .. */ style comments instead of // ... It is a NCS requeirement.

kkasperczyk-no commented on samples/matter/thermostat/src/temperature_manager.cpp:
```diff
@@ -0,0 +1,134 @@
+uint8_t ReturnReminderValue(int16_t Value)
+{
+
+    uint8_t remainder = (Value%100 + 5)/10; // Converts and returns only reminder part of value to printable type. 
```
```suggestion
    uint8_t reminder = (Value%100 + 5)/10; // Converts and returns only reminder part of value to printable type. 
```

LipinskiPNordicSemi replied:
I meant remainder value in this situation. But it might be not adequate name for this value

Damian-Nordic replied:
I think the other way around "reminder" -> "remainder" everywhere (reszta z dzielenia).

kkasperczyk-no replied:
Oh, I see :smile:

kkasperczyk-no commented on samples/matter/thermostat/src/temperature_manager.cpp:
```diff
@@ -0,0 +1,134 @@
+
+    uint8_t remainder = (Value%100 + 5)/10; // Converts and returns only reminder part of value to printable type. 
+                                            // This formula rounds reminder value to one significant figure
+    return remainder;
```
There is an area for small optimization like `return (Value%100 + 5)/10;`

kkasperczyk-no commented on samples/matter/thermostat/src/temperature_manager.cpp:
```diff
@@ -0,0 +1,134 @@
+    break;
+
+    case ThermostatAttributes::SystemMode::Id: {
+        uint8_t mode = static_cast<uint8_t>(*value);
```
Redundant casting, as the `*value` is already uint8_t, right?

kkasperczyk-no commented on samples/matter/thermostat/src/temperature_manager.cpp:
```diff
@@ -0,0 +1,134 @@
+        uint8_t mode = static_cast<uint8_t>(*value);
+        if (mThermMode != mode)
+        {
+            mThermMode = mode;
```
I'm not sure, if this check is necessary. To me, this whole case could be one-liner, like `mThermMode = *value`.

kkasperczyk-no commented on samples/matter/thermostat/CMakeLists.txt:
```diff
@@ -0,0 +1,72 @@
+    src/zcl_callbacks.cpp
+    src/zap-generated/IMClusterCommandHandler.cpp
+    src/zap-generated/callback-stub.cpp
+    src/temperature_measurement/mocked_temp.cpp
```
Shouldn't we choose mocked or real file compilation based on Kconfig option?

markaj-nordic replied:
Yes, we should do it. @LipinskiPNordicSemi Any files which are unused under certain configuration must be guarded by related KConfig. As a reference please take a look at line 55.

markaj-nordic commented on samples/matter/thermostat/Kconfig:
```diff
@@ -0,0 +1,47 @@
+mainmenu "Matter Thermostat sample application"
+
+config EXTERNAL_SENSOR
+	bool "Define if the temperature is measured on board (mocked) or fro mexternal sensor"
```
```suggestion
	bool "Select if the temperature measurement is provided directly from the board or collected from external sensor"
```

markaj-nordic replied:
We may also add a `help` prompt to indicate which value means what.

markaj-nordic commented on samples/matter/thermostat/Kconfig:
```diff
@@ -0,0 +1,47 @@
+	int "Defines mocked temperature step. Use 0 for random"
+	default 0
+	help 
+	  Use temperature step to define differences between next temperature steps, if 0 then temperature is random
```
```suggestion
	  A temperature step defining differences between consecutive mocked temperature measurements, if 0 then the step is random
```

markaj-nordic commented on samples/matter/thermostat/Kconfig:
```diff
@@ -0,0 +1,47 @@
+	  Use temperature step to define differences between next temperature steps, if 0 then temperature is random
+
+config MAXIMUM_MOCKED_TEMPERATURE
+	int "Maximum value in mocked temperature range"
```
```suggestion
	int "Maximum mocked temperature value"
```

markaj-nordic commented on samples/matter/thermostat/Kconfig:
```diff
@@ -0,0 +1,47 @@
+	default 3000
+
+config MINIMUM_MOCKED_TEMPERATURE
+	int "Minimum value in mocked temperature range"
```
```suggestion
	int "Minimum mocked temperature value"
```

markaj-nordic commented on samples/matter/thermostat/Kconfig:
```diff
@@ -0,0 +1,47 @@
+	bool "If this option is set to 'true' temperature raise"
+	default y
+	help
+	  Use this option to specify if temperature will go from maximum value to minimum, or in the opposite way
```
```suggestion
	  Specify whether the mocked temperature is rising (true) or falling (false)
```

markaj-nordic commented on samples/matter/thermostat/prj.conf:
```diff
@@ -0,0 +1,47 @@
+CONFIG_USE_SEGGER_RTT=n
+
+# Temperature sensor config 
+CONFIG_EXTERNAL_SENSOR=n
```
I'd say that if we want to immediately change the default configuration of the sample, this configuration should be made default in Kconfig definitions, so that no action is required here in the first place. If the temperature ranges are determined by the spec, I suggest to keep them in the sample.

markaj-nordic commented on samples/matter/thermostat/sample.yaml:
```diff
@@ -0,0 +1,36 @@
+    - nrf5340dk_nrf5340_cpuapp
+    - nrf7002dk_nrf5340_cpuapp
+    platform_allow: nrf52840dk_nrf52840 nrf5340dk_nrf5340_cpuapp nrf7002dk_nrf5340_cpuapp
+  sample.matter.thermostat.no_dfu:
```
Please add Michał as a collaborator so that he can take a look the sample.yml as it's going to be used in the CI twister builds.

markaj-nordic replied:
@doublemis1 please take a look at the sample.yml file.

markaj-nordic commented on samples/matter/thermostat/src/temperature_manager.cpp:
```diff
@@ -0,0 +1,118 @@
+		ReturnCompleteValue(GetHeatingSetPoint()), ReturnRemainderValue(GetHeatingSetPoint()),
+		ReturnCompleteValue(GetCoolingSetPoint()), ReturnRemainderValue(GetCoolingSetPoint()));
+
+	/*TODO - add thermostat ui update if available */
```
Should we keep this TODO?

markaj-nordic commented on samples/matter/thermostat/src/temperature_measurement/mocked_temp.cpp:
```diff
@@ -0,0 +1,58 @@
+
+LOG_MODULE_DECLARE(app, CONFIG_CHIP_APP_LOG_LEVEL);
+
+constexpr uint16_t kSimulatedReadingFrequency = (60000 / kSensorTimerPeriodMs); /*Change Simulated number at each
```
```suggestion
constexpr uint16_t kSimulatedReadingFrequency = (60000 / kSensorTimerPeriodMs); /*Change simulated number every minute
```

markaj-nordic commented on samples/matter/thermostat/src/temperature_measurement/sensor.h:
```diff
@@ -0,0 +1,41 @@
+	void TemperatureMeasurment();
+
+private:
+	int16_t previousTemperature;
```
Please use `m` prefix for these members.

markaj-nordic commented on samples/matter/thermostat/src/binding_handler.cpp:
```diff
@@ -0,0 +1,146 @@
+							      OperationalDeviceProxy *peer_device)
+{
+	auto onSuccess = [](const ConcreteDataAttributePath &attributePath, const auto &dataResponse) {
+		ChipLogProgress(NotSpecified, "Read Thermostat attribute succeeds");
```
```suggestion
		ChipLogProgress(NotSpecified, "Read Thermostat attribute succeeded");
```

markaj-nordic commented on samples/matter/thermostat/src/binding_handler.cpp:
```diff
@@ -0,0 +1,146 @@
+{
+	auto onSuccess = [](const ConcreteDataAttributePath &attributePath, const auto &dataResponse) {
+		ChipLogProgress(NotSpecified, "Read Thermostat attribute succeeds");
+		int16_t ResponseValue = dataResponse.Value();
```
```suggestion
		int16_t responseValue = dataResponse.Value();
```
nitpicking the consistent naming scheme for all variables

markaj-nordic commented on samples/matter/thermostat/src/binding_handler.h:
```diff
@@ -0,0 +1,55 @@
+		chip::ClusterId clusterId;
+		bool isGroup = false;
+		bool isReadAttribute = false;
+		uint32_t args[7];
```
Please avoid magic numbers whenever possible. Also can we default-initialize all non-initialize fields in this structure with `{}`?

markaj-nordic commented on samples/matter/thermostat/src/temp_sensor_manager.cpp:
```diff
@@ -0,0 +1,65 @@
+	AppTask::Instance().PostEvent(event);
+}
+
+/* Initiate measurement, by giving RealSensor instance  function
```
Is there anything we can do to fix this ugly formatting?

markaj-nordic commented on samples/matter/thermostat/src/temp_sensor_manager.h:
```diff
@@ -0,0 +1,36 @@
+	static void SetLocalTemperature(int16_t);
+
+private:
+	template <typename T> static void StartMeasurement(T const &Sensor);
```
```suggestion
	template <typename T> static void GetMeasurement(T const &Sensor);
```

ArekBalysNordic replied:
Something hasn't worked here. Here is still StartMeasurement, while in other places is GetMeasurement.

markaj-nordic commented on samples/matter/thermostat/src/temperature_measurement/sensor.h:
```diff
@@ -0,0 +1,41 @@
+		return sSensor;
+	};
+
+	static void TemperatureMeasurment();
```
```suggestion
	static void TemperatureMeasurement();
```
The same for `MockedSensor`

markaj-nordic commented on samples/matter/thermostat/src/temp_sensor_manager.h:
```diff
@@ -0,0 +1,36 @@
+
+#include "app_event.h"
+
+#include <app-common/zap-generated/attributes/Accessors.h>
```
Please move this include to the .cpp file.

markaj-nordic commented on samples/matter/thermostat/src/temperature_manager.cpp:
```diff
@@ -0,0 +1,118 @@
+	return static_cast<uint8_t>(Value / 100);
+}
+
+/*Converts and returns only reminder part of value to printable type.
```
```suggestion
/* Converts and returns only reminder part of value to printable type.
```

markaj-nordic commented on samples/matter/thermostat/src/temperature_manager.cpp:
```diff
@@ -0,0 +1,118 @@
+}
+
+/*Converts and returns only reminder part of value to printable type.
+This formula rounds reminder value to one significant figure*/
```
```suggestion
This formula rounds reminder value to one significant figure */
```

markaj-nordic commented on samples/matter/thermostat/src/temperature_manager.cpp:
```diff
@@ -0,0 +1,118 @@
+
+void TemperatureManager::LogThermostatStatus()
+{
+	LOG_INF("Thermostat Status - M:%d T:%d,%d'C H:%d,%d'C C:%d,%d'C", GetMode(),
```
I'm afraid this log might be a bit obscure, maybe let's use the full naming here.

LipinskiPNordicSemi replied:
Already changed

markaj-nordic commented on samples/matter/thermostat/src/temperature_manager.cpp:
```diff
@@ -0,0 +1,118 @@
+{
+	switch (attributeId) {
+	case ThermostatAttributes::LocalTemperature::Id: {
+		int16_t temp = *reinterpret_cast<int16_t *>(value);
```
`static_cast` should suffice, right?

markaj-nordic commented on samples/matter/thermostat/src/temperature_manager.h:
```diff
@@ -0,0 +1,49 @@
+using namespace chip;
+
+// AppCluster Spec Table 85.
+enum {
```
```suggestion
enum : uint8_t {
```

markaj-nordic commented on samples/matter/thermostat/README.rst:
```diff
@@ -0,0 +1,280 @@
+     Releasing the button within the six-second window cancels the factory reset procedure.
+
+Button 2:
+     Prints thermostat data to the console.
```
What about nRF7002? There it also triggers the BLE advertisement when pressed longer.

markaj-nordic commented on samples/matter/thermostat/README.rst:
```diff
@@ -0,0 +1,280 @@
+     {"fabricIndex": 1, "privilege": 1, "authMode": 2, "subjects": [<Thermostat NodeId>], "targets": \
+     [{"cluster": 1026, "endpoint": <TemperatureSensor EndpointId>, "deviceType": null}]}]' <Sensor Device NodeId> 0
+
+#. Using chipt-tool command set binding to thermosta:
```
```suggestion
#. Using chip-tool command set binding to thermostat:
```

markaj-nordic reviewed (changes requested):
General comment: please put more attention on the correct spelling :)

markaj-nordic commented on samples/matter/thermostat/prj.conf:
```diff
@@ -0,0 +1,41 @@
+
+
+
+CONFIG_TEMP_NRF5=y
```
1. Too many empty lines
2. Is this config really used now? I guess not. Anyway, even if we support real sensor measurement on board, this will not work on nRF53 as far as I know.

markaj-nordic commented on samples/matter/light_switch/README.rst:
```diff
@@ -91,14 +91,19 @@ This allows the light switch more than one lighting devices at the same time.
 In this sample, the light switch controls one or more lighting devices, but does not know the remote endpoints of the lights (on remote nodes).
 Using binding, the light switch device updates its Binding cluster with all relevant information about the lighting devices, such as their IPv6 address, node ID, and the IDs of the remote endpoints that contains the On/Off cluster and the LevelControl cluster, respectively.
 
+
```
Let's keep the original spacing here.

markaj-nordic commented on samples/matter/thermostat/src/binding_handler.h:
```diff
@@ -0,0 +1,54 @@
+/*
+ * Copyright (c) 2022 Nordic Semiconductor ASA
```
Please double check date in all added files.

ArekBalysNordic commented on samples/matter/lock/src/app_task.cpp:
```diff
@@ -209,7 +209,7 @@ CHIP_ERROR AppTask::Init()
 	GetNUSService().RegisterCommand("Lock", sizeof("Lock"), NUSLockCallback, nullptr);
 	GetNUSService().RegisterCommand("Unlock", sizeof("Unlock"), NUSUnlockCallback, nullptr);
-	if(!GetNUSService().StartServer()){
+	if (!GetNUSService().StartServer()) {
```
If you don't need to edit other projects than Thermostat please do not edit it, because the commit can become unnecessarily big.

ArekBalysNordic commented on samples/matter/thermostat/Kconfig:
```diff
@@ -0,0 +1,49 @@
+mainmenu "Matter Thermostat sample application"
+
+config EXTERNAL_SENSOR
+	bool "Select if the temperature measurement is provided directly from the board or collected from external sensor"
```
IMO config descriptions are too big. Remember, the interface that displays them probably will cut it. Please check how it looks like using: https://docs.zephyrproject.org/latest/build/kconfig/menuconfig.html

ArekBalysNordic commented on samples/matter/thermostat/prj_no_dfu.conf:
```diff
@@ -0,0 +1,44 @@
+CONFIG_CHIP=y
+CONFIG_CHIP_ENABLE_PAIRING_AUTOSTART=y
+CONFIG_CHIP_PROJECT_CONFIG="src/chip_project_config.h"
+# 32782 == 0x8000 (example Product ID added temporaly,
```
The PID of the Thermostat is set wrongly -> Please see the link below and set the proper value.

ArekBalysNordic commented on samples/matter/thermostat/prj_release.conf:
```diff
@@ -0,0 +1,47 @@
+CONFIG_CHIP=y
+CONFIG_CHIP_ENABLE_PAIRING_AUTOSTART=y
+CONFIG_CHIP_PROJECT_CONFIG="src/chip_project_config.h"
+# 32782 == 0x8000 (example Product ID added temporaly,
```
The PID of the Thermostat is set wrongly -> Please see the link below and set the proper value.

ArekBalysNordic replied:
Moreover, the note with the link below has been created only for the Template sample because in Matter there is no Template PID. In the case of the Thermostat device type, we can change it to the proper value and remove the note.

ArekBalysNordic commented on samples/matter/thermostat/src/temperature_measurement/sensor.h:
```diff
@@ -0,0 +1,41 @@
+
+#pragma once
+
+class MockedSensor {
+public:
+	static MockedSensor &Instance()
+	{
+		static MockedSensor sSensor;
+		return sSensor;
+	};
+
+	void TemperatureMeasurement();
+
+private:
+	int16_t mPreviousTemperature;
+	uint8_t mMockTempIndex;
+	uint8_t mCycleCounter;
+};
+
+class RealSensor {
+public:
+	RealSensor();
+
+	static RealSensor &Instance()
+	{
+		static RealSensor sSensor;
+		return sSensor;
+	};
+
+	static void TemperatureMeasurement();
+};
```
Hmm, I'm still wondering whether we need these two separate sensor classes here. It looks much better now than before but maybe it is worth creating a template here.

markaj-nordic replied:
I'd say it's ok. We have two different implementations of the same "interface". This gives more modularization and possibly makes it clear for the customers that they can provide their own sensor implementation.

ArekBalysNordic commented on samples/matter/thermostat/Kconfig:
```diff
@@ -0,0 +1,49 @@
+	help
+	 Set to true to enable the temperature measurement from phisical external sensor, false to mock temperature measurement
+
+config TEMPERATURE_STEP
```
As per my previous comment, I'm not sure that we should use generic config names instead of adding some prefixes, such as THERMOSTAT...

ArekBalysNordic commented on samples/matter/thermostat/src/temperature_measurement/mocked_temp.cpp:
```diff
@@ -0,0 +1,57 @@
+	mPreviousTemperature = temperature;
+
+	TempSensorManager::Instance().SetLocalTemperature(temperature);
+}
```
Missing an empty line.

ArekBalysNordic commented on samples/matter/thermostat/src/temperature_manager.h:
```diff
@@ -0,0 +1,49 @@
+using namespace chip;
+
+// AppCluster Spec Table 85.
+enum : uint8_t {
```
I guess the enum like this one should be defined somewhere in the Matter stack / Matter clusters. Could you try to find it and maybe you will avoid a duplicate.

ArekBalysNordic replied:
Moreover, I've checked the spec and I can see more than 5 entries there. What about the others? :)

ArekBalysNordic commented on samples/matter/thermostat/src/temperature_manager.cpp:
```diff
@@ -0,0 +1,121 @@
+		LOG_INF("Heating TEMP %d", mHeatingCelsiusSetPoint);
+	} break;
+
+	case ThermostatAttributes::SystemMode::Id: {
```
Wait. You're using `ThermostatAttributes::SystemMode` but you have defined your own enum in the `temperature_manager.h` file. Which one of those two SystemMode enums are you using?

ArekBalysNordic commented on samples/matter/thermostat/src/temperature_manager.cpp:
```diff
@@ -0,0 +1,121 @@
+
+constexpr EndpointId kThermostatEndpoint = 1;
+
+namespace ThermostatAttributes = chip::app::Clusters::Thermostat::Attributes;
```
What about `using chip::app::Clusters::Thermostat::Attributes` instead of creating a new namespace? But if you want to keep the code more readable by doing it this way, I'm ok with it :)

ArekBalysNordic reviewed (changes requested):
The second round of review.

markaj-nordic commented on samples/matter/thermostat/README.rst:
```diff
@@ -0,0 +1,313 @@
+
+The Access Control List (ACL) is a list related to the Access Control cluster.
+The list contains rules for managing and enforcing access control for a node's endpoints and their associated cluster instances.
+In this sample's case, this allows the temperature measurement devices to receive messages from the thermostat and run them.
```
What do you mean by "run them"? I think this needs to be rephrased.

markaj-nordic commented on samples/matter/thermostat/README.rst:
```diff
@@ -0,0 +1,313 @@
+   :start-after: matter_light_switch_sample_binding_start
+   :end-before: matter_light_switch_sample_binding_end
+
+In this sample, the thermostat can reads temperature from external temperature sensor, but does not know the remote endpoints of the lights (on remote nodes).
```
```suggestion
In this sample, the thermostat can read temperature from external temperature sensor, but does not know the remote endpoints of the lights (on remote nodes).
```
endpoints of the lights? Don't you confuse the thermostat/temperature sensor with light_switch/light_bulbs?

Anyway, I would rephrase it like:

This sample can operate in two modes:
- mocked temperature sensor mode - the thermostat application generates the simulated temperature measurements
- real temperature sensor mode - the thermostat is bound to the remote Matter temperature sensor, which provides real temperature measurements

Os something similar, later on Greg will review wording.

markaj-nordic commented on samples/matter/thermostat/README.rst:
```diff
@@ -0,0 +1,313 @@
+    :end-before: matter_door_lock_sample_led1_end
+
+Button 1:
+     Software update advertisement.
```
It's too obscure, we need to clearly describe how to interact with the UI to achieve expected behavior.

markaj-nordic commented on samples/matter/thermostat/README.rst:
```diff
@@ -0,0 +1,313 @@
+     Releasing the button within the six-second window cancels the factory reset procedure.
+
+Button 2:
+     Prints thermostat data to the console.
```
Again, it needs to be clearly stated how/low long the button should be pressed.

markaj-nordic commented on samples/matter/thermostat/README.rst:
```diff
@@ -0,0 +1,313 @@
+=======
+
+When you have built the sample and programmed it to your development kit, it automatically starts the Bluetooth LE advertising and the **LED1** starts flashing (Short Flash On).
+Thermostat should update its values every 30s and print data to logs.
```
Isn't it triggered by the button? I mean the logging.

LipinskiPNordicSemi replied:
It is updated every 30s, but you can force data update by pushing button

markaj-nordic commented on samples/matter/thermostat/README.rst:
```diff
@@ -0,0 +1,313 @@
+
+When you have built the sample and programmed it to your development kit, it automatically starts the Bluetooth LE advertising and the **LED1** starts flashing (Short Flash On).
+Thermostat should update its values every 30s and print data to logs.
+By pressing **Button 2** you can force data printing.
```
From this sentence it seems like pressing the Button 2 you start the continuous logging of the temperature. Is that so? I thought it worked like one-shot log.

markaj-nordic commented on samples/matter/thermostat/src/temp_sensor_manager.cpp:
```diff
@@ -0,0 +1,63 @@
+	PlatformMgr().UnlockChipStack();
+}
+
+template <typename T> void TempSensorManager::GetMeasurement(T const &Sensor)
```
```suggestion
template <typename T> void TempSensorManager::GetMeasurement(T const &sensor)
```
nit

LipinskiPNordicSemi committed:
samples: matter: added thermostat sample application

Created example application for thermostat device,
able to read temperature from external temperature
sensor or use mocked temperature values.

Signed-off-by: Patryk Lipinski <patryk.lipinski@nordicsemi.no>
